streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java

```diff
@@ -148,13 +205,38 @@ protected <R> QueryResult<R> runRangeQuery(final Query<R> query,
             throw new UnsupportedOperationException("Versioned stores do not support RangeQuery queries at this time.");
         }
 
+        @SuppressWarnings("unchecked")
         @Override
         protected <R> QueryResult<R> runKeyQuery(final Query<R> query,
-                                                 final PositionBound positionBound,
-                                                 final QueryConfig config) {
-            // throw exception for now to reserve the ability to implement this in the future
-            // without clashing with users' custom implementations in the meantime
-            throw new UnsupportedOperationException("Versioned stores do not support KeyQuery queries at this time.");
+            final PositionBound positionBound,
+            final QueryConfig config) {
+            if (query instanceof VersionedKeyQuery) {
+                final QueryResult<R> result;
+                final VersionedKeyQuery<K, V> typedKeyQuery = (VersionedKeyQuery<K, V>) query;
+                VersionedKeyQuery<Bytes, byte[]> rawKeyQuery;
+                if (typedKeyQuery.asOfTimestamp().isPresent()) {
+                    rawKeyQuery = VersionedKeyQuery.withKey(keyBytes(typedKeyQuery.key()));
+                    rawKeyQuery = rawKeyQuery.asOf(typedKeyQuery.asOfTimestamp().get());
+                } else {
+                    rawKeyQuery = VersionedKeyQuery.withKey(keyBytes(typedKeyQuery.key()));
+                }
+                final QueryResult<VersionedRecord<byte[]>> rawResult =
+                    wrapped().query(rawKeyQuery, positionBound, config);
+                if (rawResult.isSuccess()) {
+                    final Function<byte[], ValueAndTimestamp<V>> deserializer = getDeserializeValue(plainValueSerdes);
+                    final ValueAndTimestamp<V>  valueAndTimestamp = deserializer.apply(serializeAsBytes(rawResult.getResult()));
+                    final VersionedRecord<V> versionedRecord = new VersionedRecord<>(valueAndTimestamp.value(), valueAndTimestamp.timestamp());
+                    final QueryResult<VersionedRecord<V>> typedQueryResult =
+                        InternalQueryResultUtil.copyAndSubstituteDeserializedResult(rawResult, versionedRecord);
+                    result = (QueryResult<R>) typedQueryResult;
+                } else {
+                    // the generic type doesn't matter, since failed queries have no result set.
+                    result = (QueryResult<R>) rawResult;
+                }
+                return result;
+            }
+            // reserved for other IQv2 query types (KIP-968, KIP-969)
+            return null;
```

Comment:
Should we throw an exception instead? We should never hit this point anyway, but might be better to have a dedicate exception to guard against bugs?